decayedlog.go

```diff
@@ -0,0 +1,419 @@
+	// will store our (sharedHash, CLTV) key-value pairs.
+	defaultDbDirectory = "sharedhashes"
+
+	// sharedHashSize is the size in bytes of the keys we will be storing
+	// in the DecayedLog. It represents the first 20 bytes of a truncated
+	// sha-256 hash of a secret generated by ECDH.
+	sharedHashSize = 20
+)
+
+var (
+	// sharedHashBucket is a bucket which houses the first sharedHashSize
+	// bytes of a received HTLC's hashed shared secret as the key and the HTLC's
+	// CLTV expiry as the value.
+	sharedHashBucket = []byte("shared-hash")
+
+	// batchReplayBucket is a bucket that maps batch identifiers to
+	// serialized ReplaySets. This is used to give idempotency in the event
+	// that a batch is processed more than once.
+	batchReplayBucket = []byte("batch-replay")
+)
+
+// HashPrefix is a statically size, 20-byte array containing the prefix
+// of a Hash256, and is used to detect duplicate sphinx packets.
+type HashPrefix [sharedHashSize]byte
+
+var (
+	// ErrDecayedLogInit is used to indicate a decayed log failed to create
+	// the proper bucketing structure on startup.
+	ErrDecayedLogInit = errors.New("unable to initialize decayed log")
+
+	// ErrDecayedLogCorrupted signals that the anticipated bucketing
+	// structure has diverged since initialization.
+	ErrDecayedLogCorrupted = errors.New("decayed log structure corrupted")
+)
+
+// ReplayLog is an interface that defines a new on-disk data structure that
+// contains a persistent log to enable strong replay protection. The interface
+// is general to allow implementations near-complete autonomy. All of these
+// calls should be safe for concurrent access.
+type ReplayLog interface {
+	// Start starts up the on-disk persistent log. It returns an error if
+	// one occurs.
+	Start() error
+
+	// Stop safely stops the on-disk persistent log.
+	Stop() error
+
+	// Get retrieves an entry from the persistent log given a []byte. It
+	// returns the value stored and an error if one occurs.
+	Get([]byte) (uint32, error)
+
+	// Put stores an entry into the persistent log given a []byte and an
+	// accompanying purposefully general type. It returns an error if the
+	// provided hash prefix already exists in the log.
+	Put(*HashPrefix, uint32) error
+
+	// PutBatch stores
+	PutBatch(*Batch) (*ReplaySet, error)
+
+	// Delete deletes an entry from the persistent log given []byte
+	Delete([]byte) error
+}
+
+// DecayedLog implements the PersistLog interface. It stores the first
+// sharedHashSize bytes of a sha256-hashed shared secret along with a node's
+// CLTV value. It is a decaying log meaning there will be a garbage collector
+// to collect entries which are expired according to their stored CLTV value
+// and the current block height. DecayedLog wraps channeldb for simplicity and
+// batches writes to the database to decrease write contention.
+type DecayedLog struct {
+	started int32
+	stopped int32
+
+	dbPath string
+
+	db *channeldb.DB
```

Comment:
Why not just take *bolt.DB` instead? Still going through the PR, but it doesn't seem to use any `channeldb` specific methods.

Reply:
good call 👍 no need for extra deps